### ISSUE TITLE
Fix missing parents and commits

### ISSUE THREAD

**Issue by michitux** [kit-parco/networkit#235]
This is a follow-up to #25. Today I learned about [git-replace](https://mirrors.edge.kernel.org/pub/software/scm/git/docs/git-replace.html) which is a way to replace objects without rewriting the history. In particular, it has a special feature to replace a commit by an identical commit just with different parents. This allows to re-add all missing parents **without** breaking anything (i.e., no force pushes, broken branches etc.). As far as I understood, git also supports pushing and pulling these replacement references and while it seems they are cloned automatically, I am not sure they are pulled automatically when doing `git pull`. If we should decide to rewrite the history using `git-filter-branch`, this will persist the replacements.

Motivated by that, I got a correct mercurial import and wrote a script to get all missing parents. Doing that, I noticed that in fact not only parent pointers but actually 4 whole commits are missing. While I'm not sure yet how I'm sure it is possible to re-add them (directly with the correct parents) and then we can also reference them. Further, I would also like to check all remaining commits if they have got all parents they should have or if there are merge commits that have only one parent. To do list is thus:

- [ ] Find a way to re-add the missing commits
- [ ] Check not only commits without parents but all commits if any parents are missing
- [ ] Check that this indeed fixes the history
- [ ] Push changes to GitHub

**Comment by avdgrinten** [comment]
Hi Michael, thanks for your work. What I do not completely understand from the documentation is how replace objects are pushed/pulled. According to [some stackoverflow question](https://stackoverflow.com/questions/44028493/how-to-push-history-edited-with-git-replace-to-remote), replace objects are not pulled automatically unless you change your local `.git/config`. If that is true, we have two options:

1. Document how to pull the replace objects so that contributers who want to bisect/blame can easily pull them while other people do not need to care. This would prevent us from cleaning up the history by removing large blobs.
2. Rewrite history to make the whole history accessible to everyone.

Maybe even a third option is interesting:

3. Make a `history` branch that contains the full history in a separate repository. Squash the `Dev` branch to a single commit that is "clean" (i.e. does not contain large blobs). Add a replacement object that replaces this commit by the last commit of the `history` branch. Pro: From now on the repository is clean while the history can still be accessed and does not need to be rewritten (SHA1s remain accurate). Con: Still requires a force push to `Dev` and without pulling `history`, the history is even more shallow.

**Comment by michitux** [comment]
I also did not fully understand how push/pull works but I also assume it is not automatic. However, I got the impression that on an initial clone the replace objects are fetched so this might only affect existing clones. We can test that once I have a first version of the replace objects. Concerning the different options:

1. I would prefer this option as a first step. It does not break anything but fixes the problem for those who care. Note that this does **not** prevent us from cleaning up the history because the moment you rewrite the history using `git filter-branch` to remove large objects, the history is rewritten using the replace objects, i.e., the history is automatically fixed and persisted. If you want to use a different tool like [BFG Repo-Cleaner](https://rtyley.github.io/bfg-repo-cleaner/), you could still run `git filter-branch` to persist the fixed history first.
2. I would choose this option if we want to rewrite the history to remove large objects.
3. I do not really see the advantage of this option. I think the only advantage would be that we could keep the history of the big blobs while cleaning the history. However, this would still imply all disadvantages of the second option with the additional disadvantage that without having the `history` branch, anything forked before the `Dev` snapshot is much harder to rebase as the parent commit is missing from the new Dev branch.

**Comment by avdgrinten** [comment]
I agree that the first variant is our best option for a first step. The advantage of the last option would be that existing commit identifiers remain correct. In other words, if there are references to commits (e.g. in commit messages or issues like ("Fix a bug introduced in 12ab34cd"), parent pointers from commits in out-of-tree branches), those would still be correct once the `history` branch is pulled.

**Comment by michitux** [comment]
A short update on this: I have a python script for fixing the history that seems to be working and able to restore not just the missing parent pointers but also the 99 missing commits. I will have a closer look at the created commits on Monday and, if everything looks okay, upload them to my personal fork so we can test them.

From some more reading, it seems that replace refs are neither cloned automatically nor displayed on GitHub (see e.g. [here] (https://wiki.mozilla.org/GitHub/Gecko_Repositories)).

**Comment by michitux** [comment]
Form a short check, the new commits look good. I have just pushed them to [my fork](https://github.com/michitux/networkit/tree/Dev). Note that they are not visible on GitHub. However, you should be able to fetch them by running

```
git fetch https://github.com/michitux/networkit.git 'refs/replace/*:refs/replace/*'
```

inside your clone of NetworKit. After this download of around 150 KiB, your (local) history should be fixed, you should now have only one commit without parents and all missing commits should be there. All git commands like blame, log or show will show the new version instead of the old commits. @avdgrinten If you want you can push them to the main repository by just pushing instead of fetching the references. Then the only missing part for solution "1" would be to add some documentation how to get the references.
